subtests/docker_cli/systemd/systemd.py

```diff
@@ -0,0 +1,89 @@
+"""
+Summary
+-------
+Testing basic container as systemd service
+
+Operational Summary
+---------------------
+This is to test container as systemd service using a predefined unit file
+(``p4321.service``) and a ``script p4321-server.py``. An image is built with
+this script added. Systemd starts a container using this image & the container
+writes current time to port ``4321``. From host, the socket is read and the
+value is checked to see if it is a digit. Finally, container is stopped
+through systemd and system is cleaned up.
+
+"""
+
+import shutil
+import time
+import socket
+from os.path import exists
+from os import unlink
+from autotest.client import utils
+from autotest.client.shared.utils import is_port_free
+from dockertest.subtest import Subtest
+from dockertest.dockercmd import DockerCmd
+from dockertest.images import DockerImages
+from dockertest.output import mustpass
+from dockertest.containers import DockerContainers
+
+
+class systemd(Subtest):
+
+    def initialize(self):
+        super(systemd, self).initialize()
+        self.stuff['systemd_touchstone'] = False
+        if self.config['docker_path'] == '/usr/bin/docker':
+            shutil.copyfile(self.bindir + '/p4321-docker.service',
+                            self.config['sysd_unitf_dest'])
+        else:
+            shutil.copyfile(self.bindir + '/p4321-docker-latest.service',
+                            self.config['sysd_unitf_dest'])
+        dkrcmd = DockerCmd(self, 'build', ['--force-rm -t',
+                                           self.config['name'], self.bindir])
+        mustpass(dkrcmd.execute())
+
+    def run_once(self):
+        host = 'localhost'
+        port = 4321
+        self.sysd_action('daemon-reload')
+        self.sysd_action('start', self.config['sysd_unit_file'])
+        while is_port_free(port, host):
+            time.sleep(0.1)
+        time_from_socket = self.read_socket(host, port)
+        self.failif(not time_from_socket.isdigit(),
+                    "Data received from container is non-numeric: '%s'"
+                    % time_from_socket)
+        self.sysd_action('stop', self.config['sysd_unit_file'])
+
+    def cleanup(self):
+        super(systemd, self).cleanup()
+        cntr_name = DockerContainers(self).get_unique_name()
+        DockerImages(self).clean_all([cntr_name, self.config['name']])
+        filepath = self.config['sysd_unitf_dest']
+        if exists(filepath):
+            unlink(filepath)
+        if self.stuff['systemd_touchstone']:
+            self.sysd_action('daemon-reload')
+
+    def sysd_action(self, action=None, sysd_unit=None):
+        self.stuff['systemd_touchstone'] = True
+        command = "systemctl " + action
+        if sysd_unit:
+            command += " " + sysd_unit
```

Comment:
Minor issue:  Generally using the ``+`` token for string concatenation is frowned upon (for performance and developer cheast-beating reasons).  The "proper" way is to use ``str.format()`` or the old-style ``%`` style formatting/substitution.  Though the reasons are complicated, and personally I think for little/simple uses, the ``+`` way is more readable.  No need to change what you have.  Just FYI.